src/components/ActionButton/StatefulActionButton.ts

```diff
@@ -0,0 +1,131 @@
+     */
+    initalState: IStatefulActionButtonState;
+}
+
+/**
+ * An action button able to handle multiple states and their transitions.
+ */
+export class StatefulActionButton {
+    static ID = 'StatefulActionButton';
+    private currentState: IStatefulActionButtonState;
+    private innerActionButton: ActionButton;
+
+    constructor(public element: HTMLElement, public options: IStatefulActionButtonOptions, public bindings?: IResultsComponentBindings) {
+        if (!this.areOptionsValid()) {
+            return;
+        }
+        this.currentState = this.options.initalState;
+        this.innerActionButton = new ActionButton(element, { ...this.options.initalState, click: this.handleClick.bind(this) }, bindings);
+    }
+
+    /**
+     * Switch the state of the instance if the state and the transition between the current and new state are allowed.
+     * @param state a state to try to switch to
+     */
+    public switchTo(state: IStatefulActionButtonState) {
+        if (!this.isTransitionAllowed(state)) {
+            console.warn('This state is not allowed on this StatefulActionButton.');
+            return;
+        }
+        this.currentState.onStateExit?.apply(this);
+        state.onStateEntry?.apply(this);
+        this.innerActionButton.updateIcon(state.icon);
+        this.innerActionButton.updateTooltip(state.tooltip);
+        this.currentState = state;
+    }
+
+    /**
+     * Return the current state of the instance.
+     */
+    public getCurrentState() {
+        return this.currentState;
+    }
+
+    /**
+     * Check if the options given to the constructor are valid.
+     * If not, it will also display the appropriate warning.
+     */
+    private areOptionsValid(): boolean {
+        if (!this.options.states?.length) {
+            console.warn('The stateful action button cannot render if no states are defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (!this.options.initalState) {
+            console.warn('The stateful action button cannot render if the inital states is not defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (this.options.states.indexOf(this.options.initalState) < 0) {
+            console.warn('The stateful action button cannot render if the initial state is not in the list of states.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Check if a transition from the current state to @param state is allowed.
+     * @param state the destination of the transition
+     */
+    private isTransitionAllowed(state: IStatefulActionButtonState) {
+        return (
+            !this.options.allowedTransitions ||
+            this.options.allowedTransitions.some((transition) => transition.from === this.currentState && transition.to === state)
```

Comment:
I am not sure about relying on object references here for the comparisons. If the concrete component is not implemented exactly as expected, the comparison will fail, even for objects that look exactly the same.

Reply:
I understand your point and hesitated with id and whatnot, but preferred to go the strict way by using object references.
My main concern was that I don't want to deal with custom equality checkers, nor to deal with the function (`onStateEntry` or `onStateExit`) checks.

Reply:
I guess it might not be a problem to compare states using a state ID after all. Theoretically, one could alter the states to mess with the button behavior. One could do the same thing against the actual code.

On the good side, comparing IDs instead of references would make it easier to use.

Reply:
> Theoretically, one could alter the states to mess with the button behavior. One could do the same thing against the actual code.

Of course yes, one could break the encapsulation, but then he also breaks the implicit contract that offers a strongly typed library and he alone should deal with the pandemonium that he may face.
This is why states and transitions are specified on the constructor and are not exposed anywhere else.

The 'stringent' aspect of using object reference is my goal and I want to use 'statics' or force the reuse of the same instance to ensure no cross-contamination can occur and so the user of the button is absolutely certain of what he does when he calls `switchTo`.
Also, it's neat because most IDE will allow to go back easily to the declaration of the state, allowing you to see quite easily what are the peculiarity of your state.

Reply:
I get your point. The part that worries me is that an implementer relying on `StatefulActionButton` has to know how the states are compared internally by `StatefulActionButton` in order to correctly implement a custom button.

I think that having a map of the states and then referring to the states by IDs in the transitions would achieve the same goal. In fact, `StatefulActionButton` would perform a lookup in the map when switching from state to state. So, in the end, it would reuse the same state instances. But I think it would be less error prone for the implementer.